app/src/main/java/agency/five/codebase/android/movieapp/data/di/DataModule.kt

```diff
@@ -1,12 +1,35 @@
 package agency.five.codebase.android.movieapp.data.di
 
-import agency.five.codebase.android.movieapp.data.repository.FakeMovieRepository
+import agency.five.codebase.android.movieapp.data.database.FavoriteMovieDao
+import agency.five.codebase.android.movieapp.data.database.MovieAppDatabase
 import agency.five.codebase.android.movieapp.data.repository.MovieRepository
+import agency.five.codebase.android.movieapp.data.repository.MovieRepositoryImpl
+import androidx.room.Room
 import kotlinx.coroutines.Dispatchers
+import org.koin.android.ext.koin.androidApplication
 import org.koin.dsl.module
 
+private const val APP_DATABASE_NAME = "app_database.db"
+
 val dataModule = module {
     single<MovieRepository>{
-        FakeMovieRepository(ioDispatcher = Dispatchers.IO)
+        MovieRepositoryImpl(get(),get(),Dispatchers.IO)
+    }
+}
+
+val databaseModule = module {
+    single{
+        Room.databaseBuilder(
+            androidApplication(),
+            MovieAppDatabase::class.java,
+            APP_DATABASE_NAME
+        ).build()
+    }
+
+    fun provideFavoriteMovieDao(db: MovieAppDatabase):FavoriteMovieDao = db.getMovieDao()
```

Comment:
I suggest you inline this. Functions declaration inside lambda isn't something that should be avoided.

Reply:
Can you please be a bit more specific? I understand it's used for better optimization with higher order functions, but I'm not sure what to do with it here. I can't just put ` **inline** fun provideFavoriteMovieDao(db: MovieAppDatabase):FavoriteMovieDao = db.getMovieDao()` .

Reply:
I suggest that you remove the function and just put the code of the function on call site. 
So instead of : 
```
single{
        provideFavoriteMovieDao(get())
    }
```

You would have 
```
single{
        provideFavoriteMovieDao(get().getMovieDao())
    }
```

You can leave it as is if you like it but I usually avoid declaring functions inside another functions, because to use them they have to be declared before they are used. Also, function is supposed to do one thing (SRP), but since this function just calls another function it is a bit redundant, it doesn't extract any logic. 
Hope this makes more sense, if you wish we could further discuss this.